lib/phoenix/verified_routes.ex

```diff
@@ -0,0 +1,757 @@
+    |> Kernel.||(build_conn_forward_path(conn, router, path))
+    |> Kernel.||(path_with_script(path, conn.script_name))
+  end
+
+  def unverified_path(%URI{} = uri, _router, path) do
+    (uri.path || "") <> path
+  end
+
+  def unverified_path(%_{endpoint: endpoint}, router, path) do
+    unverified_path(endpoint, router, path)
+  end
+
+  def unverified_path(endpoint, _router, path) when is_atom(endpoint) do
+    endpoint.path(path)
+  end
+
+  def unverified_path(other, router, path) do
+    raise ArgumentError,
+          "expected a %Plug.Conn{}, a %Phoenix.Socket{}, a %URI{}, a struct with an :endpoint key, " <>
+            "or a Phoenix.Endpoint when building path for #{inspect(router)} at #{path}, got: #{inspect(other)}"
+  end
+
+  @doc false
+  def __encode_segment__(data) do
+    case data do
+      [] -> ""
+      [str | _] when is_binary(str) -> Enum.map_join(data, "/", &encode_segment/1)
+      _ -> encode_segment(data)
+    end
+  end
+
+  defp encode_segment(data) do
+    data
+    |> Phoenix.Param.to_param()
+    |> URI.encode(&URI.char_unreserved?/1)
+  end
+
+  defp verify_segment(["/" | rest], route, acc), do: verify_segment(rest, route, ["/" | acc])
+
+  # we've found a static segment, return to caller with rewritten query if found
+  defp verify_segment(["/" <> _ = segment | rest], route, acc) do
+    case {String.split(segment, "?"), rest} do
+      {[segment], _} ->
+        verify_segment(rest, route, [URI.encode(segment) | acc])
+
+      {[segment, static_query], dynamic_query} ->
+        {Enum.reverse([URI.encode(segment) | acc]),
+         verify_query(dynamic_query, route, [static_query, "?"])}
+    end
+  end
+
+  # we reached the static query string, return to caller
+  defp verify_segment(["?" <> _ = query], _route, acc) do
+    {Enum.reverse(acc), [query]}
+  end
+
+  # we reached the dynamic query string, return to call with rewritten query
+  defp verify_segment(["?" | rest], route, acc) do
+    {Enum.reverse(acc), verify_query(rest, route, ["?"])}
+  end
+
+  defp verify_segment([segment | _], route, _acc) when is_binary(segment) do
+    raise ArgumentError,
+          "path segments must begin with /, got: #{inspect(segment)} in #{Macro.to_string(route)}"
+  end
+
+  defp verify_segment(
+         [
+           {:"::", m1, [{{:., m2, [Kernel, :to_string]}, m3, [dynamic]}, {:binary, _, _} = bin]}
+           | rest
+         ],
+         route,
+         acc
+       ) do
+    unless is_binary(Enum.at(acc, 0)) do
+      raise ArgumentError,
+            "a dynamic ~p interpolation must be followed by a static segment, got: #{Macro.to_string(route)}"
+    end
+
+    rewrite = {:"::", m1, [{{:., m2, [__MODULE__, :__encode_segment__]}, m3, [dynamic]}, bin]}
+    verify_segment(rest, route, [rewrite | acc])
+  end
+
+  defp verify_segment([_other | _], route, _acc) do
+    raise ArgumentError,
+          "verified routes require a compile-time string, got: #{Macro.to_string(route)}"
+  end
```

Comment:
I don't think this last clause is ever used. In any case, we can move things to a guard and reuse it:

```suggestion
  defp verify_segment(
         [
           {:"::", m1, [{{:., m2, [Kernel, :to_string]}, m3, [dynamic]}, {:binary, _, _} = bin]}
           | rest
         ],
         route,
         [prev | _] = acc
       ) when is_binary(prev) do
    rewrite = {:"::", m1, [{{:., m2, [__MODULE__, :__encode_segment__]}, m3, [dynamic]}, bin]}
    verify_segment(rest, route, [rewrite | acc])
  end

  defp verify_segment([_ | _], route, _acc) do
    raise ArgumentError,
          "a dynamic ~p interpolation must follow a static segment, got: #{Macro.to_string(route)}"
  end
```